### ISSUE TITLE
Clarify purpose of duplicate precipitation inputs

### ISSUE THREAD

**Issue by jameshalgren** [NOAA-OWP/ngen#509]
The [CSV input files](https://github.com/NOAA-OWP/ngen/blob/master/data/forcing/cat-52_2015-12-01%2000_00_00_2015-12-30%2023_00_00.csv) include both `APCP_surface` and `precip_rate`. It would seem these are redundant. The former would be expected to be an accumulated precipitation amount and the latter a rate, and, if the are derived from the same source, should be easily transformed one to the other. But... that doesn't seem to be the case -- there is no obvious linear transformation relating the two values. (We recognize that kg/m2/s is equivalent to mm/s... it isn't a strict unit conversion issue we're seeing here, at least it doesn't appear to be...)

What is the meaning of each variable and where are the values used?

Doing some digging, it appears that `precip_rate` is used in the LSTM model and `APCP_surface` may be used in the CFE-based simulations. 

Possibly useful line references:
https://github.com/NOAA-OWP/ngen/blob/f914038ff33ab26fcf6172cf78f6bf9481706a02/include/forcing/AorcForcing.hpp#L19

https://github.com/NOAA-OWP/ngen/blob/f914038ff33ab26fcf6172cf78f6bf9481706a02/src/models/lstm/LSTM.cpp#L170

https://github.com/NOAA-OWP/ngen/blob/f914038ff33ab26fcf6172cf78f6bf9481706a02/models/tshirt/include/tshirt_c.h#L53

**Comment by jmframe** [comment]
The example LSTM models were trained with NLDAS hourly forcings: "precipitation hourly total (kg/m^2)", more information here: https://ldas.gsfc.nasa.gov/nldas/v2/forcing. However, there may have been some trickery to get one of those two variables to play nice. Looks like @madMatchstick and Scott have made some adjustments since (https://github.com/NOAA-OWP/lstm/blob/d8100a9d4f651d65e81e2e9902f9a3728e9d1049/lstm/bmi_lstm.py#L102). It would be good to re-train a model with the AORC forcing data, so it is consistent.

**Comment by hellkite500** [comment]
To clear the air on this once more, I'll explain the derivation of `precip_rate` from `APCP_surface`.

`APCP_surface` is the AORC variable in units of `kg/m2`.  This is often ASSUMED to be ~= `mm/hr`, but that is an assumption that depends ultimately on the density of water.

the `precip_rate` variable was used in some of the early model testing as an actual rate, in units of `m/s` valid until the next entry (hour, in this case) in the time series.

The exact transformation was done via these two functions.
## rho approximate density of water at temperature
This function does not consider pressure.
```python
def rho(temp): 
    """
        Calculate water density at temperature
    """
    return 999.99399 + 0.04216485*temp - 0.007097451*(temp**2) + 0.00003509571*(temp**3) - 9.9037785E-8*(temp**4) 
```
# Cumulative to rate transform transformation
This function computes the RATE at which precipitation occurs, effectively converting `kg/m^2` to `m/s` by using the AORC interval of 1 hour between inputs (the shift) in seconds, and `rho`.

```python
def aorc_as_rate(dataFrame):
    """
        Convert kg/m^2 -> m/s
    """
    if isinstance(dataFrame.index, pd.MultiIndex):
        interval = pd.Series(dataFrame.index.get_level_values(0))
    else:
        interval = pd.Series(dataFrame.index)
    interval = ( interval.shift(-1) - interval ) / np.timedelta64(1, 's')
    interval.index = dataFrame.index
    precip_rate = ( dataFrame['APCP_surface'].shift(-1) / dataFrame['TMP_2maboveground'].apply(rho) ) / interval
    precip_rate.name = 'precip_rate'
    return precip_rate
```

It is left to the user to decide which of these inputs is appropriate for a given model formulation, or whether or not the assumption that `kg/m^2` == `mm/hr` is a reasonable one for their application.

**Comment by mattw-nws** [comment]
As for the value mapping between the two being unclear, if the above explanation was a little more TL;DR than you were looking for, I'll add this briefer but less completely accurate explanation:

precip_rate[t-1] = ( ( APCP_surface[t] / 3600 ) / 1000 ) * density_factor_very_near_one

where:
* 3600 = seconds in one hour
* 1000 = millimeters in a meter
* t-1 means the previous timestep because the accumulation in APCP_surface is measured at the *end* of the hour, so the precip_rate is the rate of precip that was happening in the previous hour.

That last bit may depend on the interpretation of precip_rate: is it the rate happening in that timestep hour, or is is the rate that led to the accumulation of precip we measured at _this_ instantaneous hour marker?

In either case, I'm guessing that the bump from one timestep to the next was what was causing @jameshalgren to be completely lost as to the relationship between the two columns.

**Comment by jameshalgren** [comment]
@jmframe @mattw-nws 
Good points regarding the duplication -- my original question was about the input files. We can log a todo out of this issue to either clean up the input files or clarify the documentation or both. 

> density_factor_very_near_one
Using the equation above to compute the density factor, the temperature is in degrees Celcius (so the Kelvin value minus 273.15) 

> is it the rate happening in that timestep hour, or is is the rate that led to the accumulation of precip we measured at this instantaneous hour marker?
The same question about average temperature and instantaneous temperature applies -- which value is applicable to perform the density calculation? (though the answer is nearly inconsequential as the density difference from one hour to the next, even in cases of extraordinary temperature gradients, is less than a percent -- 0 degrees C to 40 degrees C yields 0.8% density change...)

Here's my spreadsheet, built from `cat-52_2015-12-01 00_00_00_2015-12-30 23_00_00` to see if I understood: 
![image](https://user-images.githubusercontent.com/53343824/231291916-d00db30d-eebb-44c9-a9e9-9f28a9d048ae.png)

Thanks for the input.

**Comment by jameshalgren** [comment]
Self closing. We might make the issue mentioned above: 
- [ ] Create issue regarding simplification of in-repo input files to retain only one value of precipitation, with in-line conversion capability ported to a BMI or other function.

**Comment by mattw-nws** [comment]
> Using the equation above to compute the density factor, **the temperature must be in degrees Celsius**...

Yep, this is where https://github.com/NOAA-OWP/ngen/pull/389 came from! 😆

**Comment by jameshalgren** [comment]
@joshcu
Also probably related to:
https://github.com/NOAA-OWP/cfe/issues/29
https://github.com/NOAA-OWP/ngen/issues/302

and should probably update this
https://github.com/NOAA-OWP/ngen/issues/302

**Comment by jameshalgren** [comment]
@lcunha011 and @hellkite500 
Can you clarify: 
1) Is the "AORC" dataset referred to here the Retrospective forcing for the National Water Model? 
    (i.e., files that may be browsed here:
        https://noaa-nwm-retrospective-3-0-pds.s3.amazonaws.com/index.html#CONUS/netcdf/FORCING/ 
        e.g., https://noaa-nwm-retrospective-3-0-pds.s3.amazonaws.com/CONUS/netcdf/FORCING/1998/199811250700.LDASIN_DOMAIN1). 
Or is it the AORC? (i.e., files that may be browsed through one of the following links: 
    https://registry.opendata.aws/noaa-nws-aorc/
    https://aws.amazon.com/marketplace/pp/prodview-m2sp7gsk5ts6s#resources
(it looks like it is the retrospective...)
2) Our assumption is that the `RAINRATE` variable in the retrospective datset was prepared using the algorithm given in #509 using the `APCP_surface` and `T2D` values in the AORC. Is this true? Has anyone checked that conversion? (When we compare the values in a spot check, it feels like the difference is not regular and is greater than would be expected using the posted equation.)

For #2, there could be a similar question regarding the calculation/connection between the two values in the operational 
responding here to the open issue, but cross-posting in closed issue #509, which has much more information. 
@joshcu @jmframe @castronova @igarousi -- hoping for your thoughts on this, too.
